### ISSUE TITLE
[New Resource]: aws_sagemaker_inference_component

### ISSUE THREAD

**Issue by webbermt** [hashicorp/terraform-provider-aws#35226]
### Description

Please add a new resource to manage SageMaker inference components.

### Requested Resource(s) and/or Data Source(s)

aws_sagemaker_inference_component

### Potential Terraform Configuration

```terraform
resource "aws_sagemaker_inference_component" "ic" {
  name = "my-inference-component"
  endpoint_name = aws_sagemaker_endpoint.example.name
  variant_name = "AllTraffic"

  runtime_config {
    copy_count = 1
  }
  
  specification {
    model_name = aws_sagemaker_model.example.name

    compute_resource_requirements {
      max_memory_required_in_mb = 32768
      min_memory_required_in_mb = 16384
      number_of_accelerator_devices_required = 1
      number_of_cpu_cores_required = 2
    }
    
    container {
      artifact_url = "s3://bucket/foo.tar.gz"
      image = "1234567890123.dkr.ecr.us-east-1.amazonaws.com/foo:latest"
      environment = {
        ENV_VAR1 = "foo"
        ENV_VAR2 = "foo2"
      }
    }

    startup_parameters {
      container_startup_health_check_timeout_in_seconds = 60
      model_data_download_timeout_in_seconds = 60
    }
  }

  tags = {
    Name = "foo"
  }
}
```

### References

[AWS Documentation](https://docs.aws.amazon.com/sagemaker/latest/dg/realtime-endpoints-deploy-models.html)
[AWS SageMaker CreateInferenceComponent API](https://docs.aws.amazon.com/sagemaker/latest/APIReference/API_CreateInferenceComponent.html)
[AWS Go SDK v2 SageMaker CreateInterfaceComponent](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/sagemaker#Client.CreateInferenceComponent)

### Would you like to implement a fix?

No

**Comment by isaac-smothers** [comment]
Note - inference components won't be usable until we get an optional model_name var in endpoint configurations. https://github.com/hashicorp/terraform-provider-aws/issues/40644

**Comment by fahran-wallace** [comment]
Some information to help prioritisation: Inference components are a prerequisite for some SageMaker features, including the new "[Scale down to zero](https://aws.amazon.com/about-aws/whats-new/2024/11/amazon-sagemaker-scale-down-zero-ai-inference-save-costs/)" feature. We'd like to make use of this to enable seamless disaster recovery, without wastefully leaving an instance always running in the failover region.

**Comment by fahran-wallace** [comment]
Just leaving some notes for future travellers for the workaround I attempted. I didn't get it working, but had to go quite far down the rabbithole before it became obvious it was a dead end!

TL;DR: You can't work around this by using the awscc terraform provider (at the moment).

SageMaker Inference Components _are_ available in the [AWS Cloud Control terraform provider](https://registry.terraform.io/providers/hashicorp/awscc/latest/docs/resources/sagemaker_inference_component).

I set up the awscc provider alongside the aws provider [using the docs].(https://registry.terraform.io/providers/hashicorp/aws/latest/docs/guides/using-aws-with-awscc-provider). My plan was to keep everything else provisioned using the standard provider, but use awscc for the inference component definition.

However, as @isaac-smothers mentions above (#40644), `model_name` needs to become optional in `aws_sagemaker_endpoint_configuration` before they'll be compatible with inference components. This [re:Post thread](https://repost.aws/questions/QUAioPYpWKTkCj5feSzNroxQ/create-sagemaker-endpoint-that-supports-inference-components) approximately covers it: Inference components need to be applied to an empty endpoint (that has no models defined in its endpoint configuration). Because `model_name` is mandatory, the provider currently cannot create an empty endpoint. This means the endpoint configuration resources defined by this provider can't be used in conjunction with inference components. 

Where I finally got stuck: I thought I'd be able to use the awscc provider to define the endpoint configuration as well. However, endpoint configuration isn't a supported resource type in that provider. Investigating using the [awscc's FAQ](https://github.com/hashicorp/terraform-provider-awscc/blob/main/contributing/FAQ.md#cloudformation-seems-to-support-the-resource-i-want-to-use-but-i-dont-see-it-in-the-provider), I found that it has type `NON_PROVISIONABLE`, and the awscc provider only supports resources that are `FULLY_MUTABLE`.

Once #40644 has been merged, this approach using both providers might become possible 🤞 .

**Comment by emanation** [comment]
@fahran-wallace I didn't now about awscc provider. And I'm also waiting for that feature. I was able to implement this by using a mix of terraform and boto3

```
resource "null_resource" "endpoint_configurations" {
  triggers = {
    config_name = var.config_name
  }
  provisioner "local-exec" {
    command = "python3 scripts/endpoint_config.py"
    environment = {
      CONFIG_NAME = var.config_name
      INSTANCE_TYPE = var.sagemaker_instance_type
      EXECUTION_ROLE = aws_iam_role.iam_for_sagemaker.arn
    }
  }
}

resource "aws_sagemaker_endpoint" "sagemaker_endpoint" {
  name                 = var.endpoint_name
  endpoint_config_name = var.config_name
  depends_on = [ null_resource.endpoint_configurations ]
}

resource "null_resource" "inference_component" {
  triggers = {
    endpoint_name = aws_sagemaker_endpoint.sagemaker_endpoint.name
    model_name    = aws_sagemaker_model.sagemaker_models.name
  }
  provisioner "local-exec" {
    command = "python3 scripts/interface_component.py"
    environment = {
      ENDPOINT_NAME = aws_sagemaker_endpoint.sagemaker_endpoint.name
      MODEL_NAME    = aws_sagemaker_model.sagemaker_models.name
    }
  }
  depends_on = [ aws_sagemaker_endpoint.sagemaker_endpoint, aws_sagemaker_model.sagemaker_models ]
}
```
where python scripts have `sm.create_endpoint_config` (without model) and `sm.create_inference_component` taking params from ENV vars. Also they don't return back anything thus configuration_name is a constant.
Another big caveat that it's only good for creating endpoints. It hard to updated them this way.

so, desperately waiting for this to be in next release.

**Comment by fahran-wallace** [comment]
Ooh, nice blending of terraform and boto3 @emanation ! It's unfortunate it doesn't extend to updates, but it's a great workaround for getting something working. Thanks for adding it!

**Comment by Theviki20110** [comment]
Hi there, any updates on this? Just wondering if there’s been any progress.

**Comment by RyanBaten** [comment]
FWIW the semi-ugly workaround of using terraform to spin up a cloudformation stack works for this.

**Comment by bgadzicki** [comment]
any updates on this?

**Comment by tomaszdudek7** [comment]
SageMaker Inference is hardly usable without this...

**Comment by wuodar** [comment]
Bumping
